### PR TITLE
[FEATURE][ML] Add setting for results field name

### PR DESCRIPTION
**Pull request by dimitris-athanasiou**
Adds a field in the config `dest.results_field` which
defaults to `ml`. Results will be written into an object
that is named after the results field. In addition,
upon starting the analytics there is now validation that
the source index does not already have a field matching
the results field name.

This allows composability of different analytics runs.

### CODE REVIEW

**Comment by dimitris-athanasiou** [comment]
This depends on https://github.com/elastic/ml-cpp/pull/454 to be merged in

File: x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/ml/dataframe/DataFrameAnalyticsDest.java [hunk 1]
```diff
@@ -20,40 +21,50 @@
...
+    public static final ParseField RESULTS_FIELD = new ParseField("results_field");
+
+    private static final String DEFAULT_RESULTS_FIELD = "ml";
 
     public static ConstructingObjectParser<DataFrameAnalyticsDest, Void> createParser(boolean ignoreUnknownFields) {
         ConstructingObjectParser<DataFrameAnalyticsDest, Void> parser = new ConstructingObjectParser<>("data_frame_analytics_dest",
-            ignoreUnknownFields, a -> new DataFrameAnalyticsDest((String) a[0]));
+            ignoreUnknownFields, a -> new DataFrameAnalyticsDest((String) a[0], (String) a[1]));
         parser.declareString(ConstructingObjectParser.constructorArg(), INDEX);
+        parser.declareStringOrNull(ConstructingObjectParser.optionalConstructorArg(), RESULTS_FIELD);
```
Comment (benwtrent) [inline x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/ml/dataframe/DataFrameAnalyticsDest.java]: `declareStringOrNull` I think implies that the passed argument could be explicitly `null`. This means `results_field: null` is a valid payload. This does not make sense to me. It seems to be that it being optional is good so we can set the default, but allowing it to be explicitly null and then overwriting it seems strange to me.

**Review by benwtrent — APPROVED** [review]
One minor comment about `null` values.

File: x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/ml/dataframe/DataFrameAnalyticsDest.java (hunk 1 above)
Comment (dimitris-athanasiou) [reply x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/ml/dataframe/DataFrameAnalyticsDest.java]: Ah, yes! Good spot! I got tricked to think it was the method to use for optional fields.

**Comment by droberts195** [comment]
Jenkins run elasticsearch-ci/2

**Comment by droberts195** [comment]
Jenkins run elasticsearch-ci/packaging-sample

**Comment by dimitris-athanasiou** [comment]
run elasticsearch-ci/2

**Pull request merged**